File: backend/app/schemas/vendor.py
"""
Project      : SMRITI Retail OS
Author       : Jawahar Ramkripal Mallah
Designation  : Chief Systems Architect & Creator
Email        : support@smritibooks.com
Websites     : smritibooks.com | erpnbook.com | aitdl.com
Version      : 6.16.0
Created      : 2026-09-11
Modified     : 2026-09-11
Copyright    : © SMRITIBooks.com. All Rights Reserved.
License      : Proprietary Commercial Software
Classification: Internal
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, ConfigDict, Field


class VendorStatus(str, Enum):
    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"
    BLOCKED = "BLOCKED"
    ON_HOLD = "ON_HOLD"
    PENDING_VERIFICATION = "PENDING_VERIFICATION"
    ARCHIVED = "ARCHIVED"
    MERGED = "MERGED"


class SupplierType(str, Enum):
    MANUFACTURER = "MANUFACTURER"
    DISTRIBUTOR = "DISTRIBUTOR"
    IMPORTER = "IMPORTER"
    TRADER = "TRADER"
    SERVICE_PROVIDER = "SERVICE_PROVIDER"


class CommercialClassification(str, Enum):
    PREFERRED = "PREFERRED"
    APPROVED = "APPROVED"
    CONDITIONAL = "CONDITIONAL"
    RESTRICTED = "RESTRICTED"
    BLOCKED = "BLOCKED"


class MSMECategory(str, Enum):
    MICRO = "MICRO"
    SMALL = "SMALL"
    MEDIUM = "MEDIUM"
    NOT_APPLICABLE = "NOT_APPLICABLE"


class ContactCategory(str, Enum):
    SALES = "SALES"
    ACCOUNTS = "ACCOUNTS"
    LOGISTICS = "LOGISTICS"
    MANAGEMENT = "MANAGEMENT"
    OTHER = "OTHER"
    GENERAL = "GENERAL"


# ─────────────────────────────────────────────────────────────────────────────
# Sub-Entity DTOs
# ─────────────────────────────────────────────────────────────────────────────

class VendorBankAccountDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: Optional[str] = None
    bank_name: str
    account_holder_name: str
    account_number: str
    ifsc: str
    branch: Optional[str] = None
    account_type: str = "CURRENT"
    is_primary: bool = False
    verification_status: str = "PENDING"
    verified_at: Optional[datetime] = None


class VendorContactDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: Optional[str] = None
    contact_name: str
    contact_category: str = "GENERAL"
    designation: Optional[str] = None
    department: Optional[str] = None
    phone: Optional[str] = None
    mobile: Optional[str] = None
    email: Optional[str] = None
    is_primary: bool = False


class VendorAddressDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: Optional[str] = None
    address_type: str = "BILLING"
    address_title: Optional[str] = None
    address_line1: str
    address_line2: Optional[str] = None
    city: str
    state: str
    state_code: Optional[str] = None
    pincode: str
    country: str = "India"
    gstin: Optional[str] = None
    is_primary: bool = False


class VendorCommercialProfileDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    supplier_type: str = "DISTRIBUTOR"
    payment_terms_days: int = 30
    msme_registration_no: Optional[str] = None
    msme_category: str = "NOT_APPLICABLE"
    commercial_classification: str = "APPROVED"
    tds_section: Optional[str] = "194Q"
    tds_rate: float = 0.10
    tax_treatment: str = "REGISTERED_REGULAR"
    outstanding_liability: float = 0.0


class VendorComplianceProfileDTO(BaseModel):
    gstin: Optional[str] = None
    pan: Optional[str] = None
    msme_registration_no: Optional[str] = None
    msme_category: str = "NOT_APPLICABLE"
    verification_flags: Dict[str, Any] = Field(default_factory=dict)
    gst_verified: bool = False
    pan_verified: bool = False
    bank_verified: bool = False
    msme_verified: bool = False


# ─────────────────────────────────────────────────────────────────────────────
# Canonical Vendor DTOs
# ─────────────────────────────────────────────────────────────────────────────

class VendorSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    code: str
    legal_name: str
    trade_name: Optional[str] = None
    gstin: Optional[str] = None
    pan: Optional[str] = None
    mobile: Optional[str] = None
    email: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    status: str = "ACTIVE"
    commercial_classification: str = "APPROVED"
    supplier_type: str = "DISTRIBUTOR"
    outstanding: float = 0.0


class VendorDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    code: str
    legal_name: str
    trade_name: Optional[str] = None
    party_type: str = "ORGANIZATION"
    gstin: Optional[str] = None
    pan: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    mobile: Optional[str] = None
    address_line1: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    pincode: Optional[str] = None
    status: str = "ACTIVE"
    merged_into_party_id: Optional[str] = None
    legacy_supplier_id: Optional[str] = None

    commercial: VendorCommercialProfileDTO
    compliance: VendorComplianceProfileDTO
    contacts: List[VendorContactDTO] = Field(default_factory=list)
    addresses: List[VendorAddressDTO] = Field(default_factory=list)
    bank_accounts: List[VendorBankAccountDTO] = Field(default_factory=list)
    roles: List[str] = Field(default_factory=lambda: ["SUPPLIER"])
    tags: List[str] = Field(default_factory=list)


class VendorCreateRequest(BaseModel):
    code: Optional[str] = None
    legal_name: str
    trade_name: Optional[str] = None
    party_type: str = "ORGANIZATION"
    gstin: Optional[str] = None
    pan: Optional[str] = None
    email: Optional[str] = None
    mobile: Optional[str] = None
    phone: Optional[str] = None
    status: str = "ACTIVE"

    # Primary address
    address_line1: Optional[str] = None
    address_line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    pincode: Optional[str] = None

    # Commercial & Statutory
    commercial: Optional[VendorCommercialProfileDTO] = None
    contacts: List[VendorContactDTO] = Field(default_factory=list)
    addresses: List[VendorAddressDTO] = Field(default_factory=list)
    bank_accounts: List[VendorBankAccountDTO] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)


class VendorUpdateRequest(BaseModel):
    legal_name: Optional[str] = None
    trade_name: Optional[str] = None
    gstin: Optional[str] = None
    pan: Optional[str] = None
    email: Optional[str] = None
    mobile: Optional[str] = None
    phone: Optional[str] = None
    status: Optional[str] = None
    address_line1: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    pincode: Optional[str] = None

    commercial: Optional[VendorCommercialProfileDTO] = None
    contacts: Optional[List[VendorContactDTO]] = None
    addresses: Optional[List[VendorAddressDTO]] = None
    bank_accounts: Optional[List[VendorBankAccountDTO]] = None
    tags: Optional[List[str]] = None


class VendorMergeRequest(BaseModel):
    primary_vendor_id: str
    secondary_vendor_id: str
    merge_reason: str = "DUPLICATE_CONVERGENCE"


class VendorMergeResponse(BaseModel):
    success: bool
    primary_vendor_id: str
    secondary_vendor_id: str
    message: str

File: src/types/vendor.ts
/**
 * Project      : SMRITI Retail OS
 * Author       : Jawahar Ramkripal Mallah
 * Designation  : Chief Systems Architect & Creator
 * Email        : support@smritibooks.com
 * Websites     : smritibooks.com | erpnbook.com | aitdl.com
 * Version      : 6.16.0
 * Created      : 2026-09-11
 * Modified     : 2026-09-11
 * Copyright    : © SMRITIBooks.com. All Rights Reserved.
 * License      : Proprietary Commercial Software
 * Classification: Internal
 */

export type VendorStatus = 
  | "ACTIVE"
  | "INACTIVE"
  | "BLOCKED"
  | "ON_HOLD"
  | "PENDING_VERIFICATION"
  | "ARCHIVED"
  | "MERGED";

export type SupplierType = 
  | "MANUFACTURER"
  | "DISTRIBUTOR"
  | "IMPORTER"
  | "TRADER"
  | "SERVICE_PROVIDER";

export type CommercialClassification = 
  | "PREFERRED"
  | "APPROVED"
  | "CONDITIONAL"
  | "RESTRICTED"
  | "BLOCKED";

export type MSMECategory = 
  | "MICRO"
  | "SMALL"
  | "MEDIUM"
  | "NOT_APPLICABLE";

export type ContactCategory = 
  | "SALES"
  | "ACCOUNTS"
  | "LOGISTICS"
  | "MANAGEMENT"
  | "OTHER"
  | "GENERAL";

export interface VendorBankAccount {
  id?: string;
  bankName: string;
  accountHolderName: string;
  accountNumber: string;
  ifsc: string;
  branch?: string;
  accountType: "CURRENT" | "SAVINGS" | "CC" | "OVERDRAFT";
  isPrimary: boolean;
  verificationStatus: "PENDING" | "VERIFIED" | "REJECTED";
  verifiedAt?: string;
}

export interface VendorContact {
  id?: string;
  contactName: string;
  contactCategory: ContactCategory;
  designation?: string;
  department?: string;
  phone?: string;
  mobile?: string;
  email?: string;
  isPrimary: boolean;
}

export interface VendorAddress {
  id?: string;
  addressType: "BILLING" | "SHIPPING" | "WAREHOUSE" | "REGISTERED_OFFICE" | "BRANCH";
  addressTitle?: string;
  addressLine1: string;
  addressLine2?: string;
  city: string;
  state: string;
  stateCode?: string;
  pincode: string;
  country: string;
  gstin?: string;
  isPrimary: boolean;
}

export interface VendorCommercialProfile {
  supplierType: SupplierType;
  paymentTermsDays: number;
  msmeRegistrationNo?: string;
  msmeCategory: MSMECategory;
  commercialClassification: CommercialClassification;
  tdsSection?: string;
  tdsRate: number;
  taxTreatment: string;
  outstandingLiability: number;
}

export interface VendorComplianceProfile {
  gstin?: string;
  pan?: string;
  msmeRegistrationNo?: string;
  msmeCategory: MSMECategory;
  verificationFlags: {
    gstVerified?: boolean;
    panVerified?: boolean;
    bankVerified?: boolean;
    msmeVerified?: boolean;
  };
}

export interface VendorSummary {
  id: string;
  code: string;
  legalName: string;
  tradeName?: string;
  gstin?: string;
  pan?: string;
  mobile?: string;
  email?: string;
  city?: string;
  state?: string;
  status: VendorStatus;
  commercialClassification: CommercialClassification;
  supplierType: SupplierType;
  outstanding: number;
}

export interface VendorDetail {
  id: string;
  code: string;
  legalName: string;
  tradeName?: string;
  partyType: string;
  gstin?: string;
  pan?: string;
  email?: string;
  phone?: string;
  mobile?: string;
  addressLine1?: string;
  city?: string;
  state?: string;
  pincode?: string;
  status: VendorStatus;
  mergedIntoPartyId?: string;

  commercial: VendorCommercialProfile;
  compliance: VendorComplianceProfile;
  contacts: VendorContact[];
  addresses: VendorAddress[];
  bankAccounts: VendorBankAccount[];
  roles: string[];
  tags: string[];
}

export interface VendorCreateRequest {
  code?: string;
  legalName: string;
  tradeName?: string;
  partyType?: string;
  gstin?: string;
  pan?: string;
  email?: string;
  mobile?: string;
  phone?: string;
  status?: VendorStatus;
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  state?: string;
  pincode?: string;
  commercial?: Partial<VendorCommercialProfile>;
  contacts?: VendorContact[];
  addresses?: VendorAddress[];
  bankAccounts?: VendorBankAccount[];
  tags?: string[];
}

export interface VendorUpdateRequest {
  legalName?: string;
  tradeName?: string;
  gstin?: string;
  pan?: string;
  email?: string;
  mobile?: string;
  phone?: string;
  status?: VendorStatus;
  addressLine1?: string;
  city?: string;
  state?: string;
  pincode?: string;
  commercial?: Partial<VendorCommercialProfile>;
  contacts?: VendorContact[];
  addresses?: VendorAddress[];
  bankAccounts?: VendorBankAccount[];
  tags?: string[];
}

export type PartyRole = "SUPPLIER" | "CUSTOMER" | "EMPLOYEE" | "TRANSPORTER";
export type PartyType = "ORGANIZATION" | "INDIVIDUAL";
export type AddressType = "BILLING" | "SHIPPING" | "WAREHOUSE" | "REGISTERED_OFFICE" | "BRANCH";
export type BankAccountType = "CURRENT" | "SAVINGS" | "CC" | "OVERDRAFT";
export type BankVerificationStatus = "PENDING" | "VERIFIED" | "REJECTED";
export type TaxTreatment = "REGISTERED_REGULAR" | "REGISTERED_COMPOSITION" | "UNREGISTERED" | "OVERSEAS" | "SEZ";

export type VendorCommercial = VendorCommercialProfile;
export type VendorCompliance = VendorComplianceProfile;

export interface VendorMergeRequest {
  sourcePartyId: string;
  targetPartyId: string;
  mergeReason: string;
  reassignTransactions?: boolean;
}

export interface VendorMergeResponse {
  success: boolean;
  message: string;
  sourcePartyId: string;
  targetPartyId: string;
  reassignedCounts?: Record<string, number>;
}

export type VendorTabId =
  | "overview"
  | "identity"
  | "addresses"
  | "contacts"
  | "commercial"
  | "banking"
  | "procurement"
  | "payables"
  | "scorecard";

